### ISSUE TITLE
Performing multiple measurement on Safari for iOS will exceed memory limit

### ISSUE THREAD

**Issue by frankshaka** [jhildenbiddle/canvas-size#16]
Running the following code on Safari for iOS (15+) will cause the browser to report errors:

```javascript
  try {
    const test1 = await canvasSize.maxWidth({ usePromise: true })
    console.log('MaxWidth:', test1.width)
    const test2 = await canvasSize.maxHeight({ usePromise: true })
    console.log('MaxHeight:', test2.height)
    const test3 = await canvasSize.maxArea({ usePromise: true })
    console.log('MaxArea:', test3.width, test3.height)
  } catch (err) {
    console.error('Failed to measure canvas size:', err)
  }
```

<img width="755" alt="Screen Shot 2022-10-19 at 12 57 03 PM" src="https://user-images.githubusercontent.com/1614752/196601153-853aec7b-1a38-4522-8f10-ddef09a54b21.png">

After some research, it seems like we have to "release" the canvas element created for testing immediately after the test is finished, to prevent the browser from holding the memory allocated for the canvas' storage. This affects not only all following canvas tests, but also real canvas applications that load afterwards, as long as they share the same memory usage constraint.

FYI: https://pqina.nl/blog/total-canvas-memory-use-exceeds-the-maximum-limit

**Comment by stephen-dirtcreative** [comment]
I added the following code in the canvasTest function to patch the dist files in the canvas-size package (both canvas-size.esm.js and canvas-size.js).

After the line: `var isTestPass = cropCtx && cropCtx.getImageData(0, 0, 1, 1).data[3] !== 0;`

```
     // reduce canvas size to limit Safari canvas hoarding memory issues
     if (testCtx) {
       testCtx.width = 1;
       testCtx.height = 1;
       testCtx.clearRect(0, 0, 1, 1);
     }
```

On the iOS 16 simulator I did see warnings about exceeding the canvas memory limit for the first 2 test sizes, but the test set did complete successfully with the 4096 size (max listed for iOS). I also only used the 3 most common sizes to test (I don't worry about differentiating possibly large oblong canvas dimensions, and am content with squares) to limit any performance penalty since the web worker flow won't work in Safari, either. My use case primarily is concerned with zooming PDFs via PDF.js so these values contribute to setting the effective zoom limit I can reach for a given file.:

```
  const testSizes = [
    // latest known chromium desktop value (Nov. 2022)
    16384,
    // latest known firefox desktop (Nov. 2022)
    11180,
    // latest known iOS mobile (Nov. 2022)
    4096
  ];

  // ...

  canvasSize.test({
    sizes: [...(testSizes.map((s) => [s, s]))],
    usePromise: true,
  }) //.then...
```

Finally, since I don't work in a vacuum, I use patch-package (available via npm) so that anyone else working with my repository will get the changes. I didn't make use of its ability to submit an issue as providing a patch to dist files doesn't seem particularly more useful than a descriptive response like this one.

**Comment by jhildenbiddle** [comment]
@frankshaka @stephen-dirtcreative 

Thank yo goth for your input. I'll review this later this week and publish an update assuming all goes well.

**Comment by jhildenbiddle** [comment]
@frankshaka, @stephen-dirtcreative --

I have implemented a fix and made it available on a separate branch for testing:

- Branch: https://github.com/jhildenbiddle/canvas-size/tree/fix-16
- Fix: https://github.com/jhildenbiddle/canvas-size/commit/174cdbc4a503545ae6885e73bb8866615c0a8a4f#diff-6a0b02710bf4700c3a0a66bf90074ec1b73fcad6903c54ed9d6551df9fd63f32

The fix addresses the issue in Safari 16 on macOS, but admittedly I have not tried it with iOS. If you can test this version in your application and verify that the issue is resolved, I will publish a new version as soon as possible.

FYI: According to [this comment](https://stackoverflow.com/a/56975986/4903063) on StackOverflow, Safari will continue to hold on to canvas element more aggressively when the Web Inspector is open. Something to keep in mind while you are troubleshooting.

Thanks!

**Comment by jhildenbiddle** [comment]
👆 @frankshaka @stephen-dirtcreative

**Comment by frankshaka** [comment]
> @frankshaka, @stephen-dirtcreative --
> 
> * Branch: https://github.com/jhildenbiddle/canvas-size/tree/fix-16

This fix works in my app. Great!

In addition, I tested it against the code snippet in my original post, and it also works as expected:

```javascript
  try {
    const test1 = await canvasSize.maxWidth({ usePromise: true })
    console.log('MaxWidth:', test1.width)
    const test2 = await canvasSize.maxHeight({ usePromise: true })
    console.log('MaxHeight:', test2.height)
    const test3 = await canvasSize.maxArea({ usePromise: true })
    console.log('MaxArea:', test3.width, test3.height)
  } catch (err) {
    console.error('Failed to measure canvas size:', err)
  }
```

<img width="757" alt="Screen Shot 2023-02-20 at 6 21 58 PM" src="https://user-images.githubusercontent.com/1614752/220078632-119b33a6-6d6e-45dc-bf19-6670c8c92c12.png">

The warnings reported during canvas size testing can be ignored as long as the temporary canvas storage is released properly.

**Comment by jhildenbiddle** [comment]
Thanks for the creating the issue, the tip on how to fix it, and testing/feedback, @frankshaka. Fix has been published in v1.2.6.
